numpy/_core/src/multiarray/usertypes.c

```diff
@@ -253,12 +260,54 @@ PyArray_RegisterDataType(PyArray_Descr *descr)
 
     snprintf(name, name_length, "numpy.dtype[%s]", scalar_name);
 
+    /*
+     * Copy the user provided descriptor struct into a new one.  This is done
+     * in order to allow different layout between the two.
+     */
+    PyArray_Descr *descr = PyObject_Malloc(sizeof(PyArray_Descr));
+    if (descr == NULL) {
+        PyMem_FREE(name);
+        PyErr_NoMemory();
+        return -1;
+    }
+    PyObject_INIT(descr, Py_TYPE(descr_proto));
```

Comment:
I'm super confused here - why allocate something of size `PyArray_Descr` and then initialize with the different type?

Reply:
Just thought it is more minimal/obvious that it doesn't change anything.  This could be subclasses, although we are replacing the type with a new one in any case, so I don't _think_ it can matter (but didn't check).

Reply:
I guess I was mostly confused why we don't initialize as PyArray_Descr` directly,.

Reply:
Docs will clarify it:  Existing user dtypes need a stable ABI. `PyArray_DescrProto` being identical to the old `PyArray_Descr` gives them that.
So a minor code change and `#if #define` allows you to have code that will both run and compile with 1.x and 2.x.

Reply:
But I thought the point here was to copy the old type of information into a new type...  I guess bottom line would be to add a comment here...

Reply:
Right, exactly.  With the reason being that what is passed in as `descr_proto` remains ABI compatible, but what gets registers if of course not.